documents/entity/service_test.go

```diff
@@ -0,0 +1,362 @@
+		queueSrv,
+		ctx[transactions.BootstrappedService].(transactions.Manager))
+}
+
+func TestService_Update(t *testing.T) {
+	_, srv := getServiceWithMockedLayers()
+	eSrv := srv.(service)
+	ctxh := testingconfig.CreateAccountContext(t, cfg)
+
+	// missing last version
+	model, _ := createCDWithEmbeddedEntity(t)
+	_, _, _, err := eSrv.Update(ctxh, model)
+	assert.Error(t, err)
+	assert.True(t, errors.IsOfType(documents.ErrDocumentNotFound, err))
+	assert.NoError(t, testRepo().Create(accountID, model.CurrentVersion(), model))
+
+	// calculate data root fails
+	nm := new(mockModel)
+	nm.On("ID").Return(model.ID(), nil).Once()
+	_, _, _, err = eSrv.Update(ctxh, nm)
+	nm.AssertExpectations(t)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "unknown document type")
+
+	// success
+	data, err := eSrv.DeriveEntityData(model)
+	assert.Nil(t, err)
+	data.LegalName = "test company"
+	data.Contacts = []*entitypb.Contact{{Name: "Mr. Test"}}
+	collab := testingidentity.GenerateRandomDID().String()
+	newInv, err := eSrv.DeriveFromUpdatePayload(ctxh, &cliententitypb.EntityUpdatePayload{
+		Identifier:    hexutil.Encode(model.ID()),
+		Collaborators: []string{collab},
+		Data:          data,
+	})
+	assert.Nil(t, err)
+	newData, err := eSrv.DeriveEntityData(newInv)
+	assert.Nil(t, err)
+	assert.Equal(t, data, newData)
+
+	model, _, _, err = eSrv.Update(ctxh, newInv)
+	assert.Nil(t, err)
+	assert.NotNil(t, model)
+	assert.True(t, testRepo().Exists(accountID, model.ID()))
+	assert.True(t, testRepo().Exists(accountID, model.CurrentVersion()))
+	assert.True(t, testRepo().Exists(accountID, model.PreviousVersion()))
+
+	newData, err = eSrv.DeriveEntityData(model)
+	assert.Nil(t, err)
+	assert.Equal(t, data, newData)
+}
+
+func TestService_DeriveFromUpdatePayload(t *testing.T) {
+	_, eSrv := getServiceWithMockedLayers()
+	// nil payload
+	doc, err := eSrv.DeriveFromUpdatePayload(nil, nil)
+	assert.Error(t, err)
+	assert.True(t, errors.IsOfType(documents.ErrDocumentNil, err))
+	assert.Nil(t, doc)
+
+	// nil payload data
+	doc, err = eSrv.DeriveFromUpdatePayload(nil, &cliententitypb.EntityUpdatePayload{})
+	assert.Error(t, err)
+	assert.True(t, errors.IsOfType(documents.ErrDocumentNil, err))
+	assert.Nil(t, doc)
+
+	// messed up identifier
+	contextHeader := testingconfig.CreateAccountContext(t, cfg)
+	payload := &cliententitypb.EntityUpdatePayload{Identifier: "some identifier", Data: &cliententitypb.EntityData{}}
+	doc, err = eSrv.DeriveFromUpdatePayload(contextHeader, payload)
+	assert.Error(t, err)
+	assert.True(t, errors.IsOfType(documents.ErrDocumentIdentifier, err))
+	assert.Contains(t, err.Error(), "failed to decode identifier")
+	assert.Nil(t, doc)
+
+	// missing last version
+	id := utils.RandomSlice(32)
+	payload.Identifier = hexutil.Encode(id)
+	doc, err = eSrv.DeriveFromUpdatePayload(contextHeader, payload)
+	assert.Error(t, err)
+	assert.True(t, errors.IsOfType(documents.ErrDocumentNotFound, err))
+	assert.Nil(t, doc)
+
+	// failed to load from data
```

Comment:
maybe this descriptor comment should be called `Entity Data does not contain an identity` just to be more specific